packages/medusa/src/services/cart.ts

```diff
@@ -1735,77 +1761,140 @@ class CartService extends TransactionBaseService {
+          )
+          return
+        }
+
+        const providerSet = new Set(region.payment_providers.map((p) => p.id))
+        const alreadyConsumedProviderIds: Set<string> = new Set()
+
         const partialSessionInput: Omit<PaymentSessionInput, "provider_id"> = {
           cart: cart as Cart,
           customer: cart.customer,
-          amount: cart.total,
+          amount: total,
           currency_code: cart.region.currency_code,
         }
+        const partialPaymentSessionData = {
+          cart_id: cartId,
+          data: {},
+          status: PaymentSessionStatus.PENDING,
+          amount: total,
+        }
 
-        // If there are existing payment sessions ensure that these are up to date
-        const seen: string[] = []
-        if (cart.payment_sessions?.length) {
-          await Promise.all(
-            cart.payment_sessions.map(async (paymentSession) => {
-              if (
-                total <= 0 ||
-                !region.payment_providers.find(
-                  ({ id }) => id === paymentSession.provider_id
-                )
-              ) {
-                return this.paymentProviderService_
-                  .withTransaction(transactionManager)
-                  .deleteSession(paymentSession)
-              } else {
-                seen.push(paymentSession.provider_id)
-
-                const paymentSessionInput = {
-                  ...partialSessionInput,
-                  provider_id: paymentSession.provider_id,
-                }
+        await Promise.all(
+          cart.payment_sessions.map(async (session) => {
+            if (!providerSet.has(session.provider_id)) {
+              /**
+               * if the provider does not belong to the region then delete the session.
+               * The deletion occurs locally if there is no external data or if it is not selected
+               * otherwise the deletion will also occur remotely through the external provider.
+               */
+
+              return await deleteSessionAppropriately(session)
+            }
 
-                return this.paymentProviderService_
-                  .withTransaction(transactionManager)
-                  .updateSession(paymentSession, paymentSessionInput)
+            /**
+             * if the provider belongs to the region then update or delete the session.
+             * The update occurs locally if it is not selected
+             * otherwise the update will also occur remotely through the external provider.
+             * In case the session is not selected but contains an external provider data, we delete the external provider
+             * session to be in a clean state.
+             */
+
+            // We are saving the provider id on which the work below will be done. That way,
+            // when handling the providers from the cart region at a later point below, we do not double the work on the sessions that already
+            // exists for the same provider.
```

Comment:
nit(non-blocking): Comment format should match the one above i.e. use `/** ... */`

Reply:
The above comment is a block comment, explaining the flow in a wider manner until the next block. Those are local comments to describe an action in the block. Are you sure? I am afraid that then the above block comment is not interpreted properly like describing the block below including this one. Wdyt?

Reply:
Alright, let's keep them as is 👍